src/bridge/cockpitpcpmetrics.c

```diff
@@ -0,0 +1,1017 @@
+                }
+
+              /* HACK: We can't use json_builder_add_string_value here since
+                 it turns empty strings into 'null' values inside arrays.
+
+                 https://bugzilla.gnome.org/show_bug.cgi?id=730803
+              */
+              {
+                JsonNode *string_element = json_node_alloc ();
+                json_node_init_string (string_element, instance);
+                json_array_add_element (array, string_element);
+              }
+
+              if (instance)
+                free (instance);
+            }
+          json_array_add_array_element (instances, array);
+        }
+    }
+  json_object_set_array_member (root, "instances", instances);
+
+  units = json_array_new ();
+  for (i = 0; i < result->numpmid; i++)
+    {
+      if (self->metrics[i].factor == 1.0)
+        {
+          json_array_add_string_element (units, pmUnitsStr(self->metrics[i].units));
+        }
+      else
+        {
+          gchar *name = g_strdup_printf ("%s*%g", pmUnitsStr(self->metrics[i].units), 1.0/self->metrics[i].factor);
+          json_array_add_string_element (units, name);
+          g_free (name);
+        }
+    }
+  json_object_set_array_member (root, "units", units);
+
+  types = json_array_new ();
+  for (i = 0; i < result->numpmid; i++)
+    {
+      switch (self->metrics[i].desc.type) {
+      case PM_TYPE_STRING:
+        json_array_add_string_element (types, "string");
+        break;
+      case PM_TYPE_32:
+      case PM_TYPE_U32:
+      case PM_TYPE_64:
+      case PM_TYPE_U64:
+      case PM_TYPE_FLOAT:
+      case PM_TYPE_DOUBLE:
+        json_array_add_string_element (types, "number");
+        break;
+      default:
+        json_array_add_null_element (types);
+        break;
+      }
+    }
+  json_object_set_array_member (root, "types", types);
+
+  semantics = json_array_new ();
+  for (i = 0; i < result->numpmid; i++)
+    {
+      switch (self->metrics[i].desc.sem) {
+      case PM_SEM_COUNTER:
+        json_array_add_string_element (semantics, "counter");
+        break;
+      case PM_SEM_INSTANT:
+        json_array_add_string_element (semantics, "instant");
+        break;
+      case PM_SEM_DISCRETE:
+        json_array_add_string_element (semantics, "discrete");
+        break;
+      default:
+        json_array_add_null_element (types);
+        break;
+      }
+    }
+  json_object_set_array_member (root, "semantics", semantics);
+
+  return root;
+}
+
+static JsonObject *
+build_meta_if_necessary (CockpitPcpMetrics *self,
+                         pmResult *result)
+{
+  if (self->last)
+    {
+      /*
+       * If we've already sent the first meta message, then only send
+       * another when the set of instances in the results change.
+       */
+
+      if (result_meta_equal (self->last, result))
+        return NULL;
+    }
+
+  return build_meta (self, result);
+}
+
+typedef struct {
+  JsonArray *array;
+  int n_skip;
+} CompressedArrayBuilder;
```

Comment:
This code should really be in CockpitMetrics as it will need to be used elsewhere.